tests/integration/test_charm.py

```diff
@@ -4,148 +4,158 @@
     ]
-    for query in queries:
-        cursor.execute(query)
-    cnx.commit()
-    cursor.close()
-    cnx.close()
-    # replication take a little time
-    sleep(1)
-    # test on all secondaries
-    for unit_id in UNIT_IDS[1:]:
-        secondary_host_ip = await get_pod_ip(ops_test, f"{APP_NAME}/{unit_id}")
-        cnx = db_connect(secondary_host_ip, root_password)
-        cursor = cnx.cursor()
-        # Query the table
-        cursor.execute("SELECT * FROM mysql.charmtest;")
-        # fetch the result
-        result = cursor.fetchall()
-        assert result[0][0] == "hello"
-
-
-async def get_password(ops_test: OpsTest, password_key: str) -> str:
-    """Get password using the action."""
-    unit = ops_test.model.units.get(f"{APP_NAME}/0")
-    action = await unit.run_action("get-root-credentials")
-    result = await action.wait()
-    return result.results[password_key]
-
-
-def db_connect(host: str, root_password: str):
-    """Create a connection to the MySQL server.
-
-    uses root user to connect to the MySQL server.
-
-    Returns:
-        MySQLConnection: connection to the MySQL server
-    """
-    cnx = mysql.connector.connect(user="root", password=root_password, host=host)
-    return cnx
-
-
-async def get_pod_ip(ops_test: OpsTest, unit_name: str) -> str:
-    """Get the pod IP address for the given unit.
-
-    Args:
-        ops_test: test fixture
-        unit_name: unit name
-    Returns:
-        str: pod IP address
-    """
-    status = await ops_test.model.get_status()
-    for _, unit in status["applications"][APP_NAME]["units"].items():
-        if unit["provider-id"] == unit_name.replace("/", "-"):
-            pod_ip = unit["address"]
-            break
-
-    return pod_ip
+
+    # Confirm that the values are available on all units
+    for unit in ops_test.model.applications[APP_NAME].units:
+        unit_address = await get_unit_address(ops_test, unit.name)
+
+        output = await execute_queries_on_unit(
+            unit_address,
+            server_config_credentials["username"],
+            server_config_credentials["password"],
+            select_data_sql,
+        )
+        assert random_chars in output
+
+
+@pytest.mark.order(3)
+@pytest.mark.abort_on_fail
+async def test_scale_up_and_down(ops_test: OpsTest) -> None:
+    """Confirm that a new primary is elected when the current primary is torn down."""
+    async with ops_test.fast_forward():
+        random_unit = ops_test.model.applications[APP_NAME].units[0]
+
+        await scale_application(ops_test, APP_NAME, 5)
+
+        cluster_status = await get_cluster_status(ops_test, random_unit)
+        online_member_addresses = [
+            member["address"]
+            for _, member in cluster_status["defaultreplicaset"]["topology"].items()
+            if member["status"] == "online"
+        ]
+        assert len(online_member_addresses) == 5
+
+        await scale_application(ops_test, APP_NAME, 1)
```

Comment:
noob question, doesn't this not always tear down the primary, ie the priamry could have been the single unit remaining? (ie primary removal will get tested 4/5 of the time?)

Reply:
the integration test is not actually testing for the scale down of the primary - instead testing how the loss of quorum is dealt with by the charm when scaling down (when going from 5 online nodes to 1 online node)